File: XamarinStore.Droid/Fragments/BasketFragment.cs
using System;
using System.Collections.Generic;
using System.Globalization;

using System.Linq;
using System.Text;
using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Util;
using Android.Views;
using Android.Widget;

namespace XamarinStore
{
	public class BasketFragment : ListFragment
	{
		Order _order;
		Button checkoutButton;

		public BasketFragment ()
		{

		}

		public BasketFragment(Order order)
		{
			_order = order;
		}

		public override void OnCreate (Bundle savedInstanceState)
		{
			base.OnCreate (savedInstanceState);
			RetainInstance = true;
		}

		public override View OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
		{
			var shoppingCartView = inflater.Inflate (Resource.Layout.Basket, container, false);

			checkoutButton = shoppingCartView.FindViewById<Button> (Resource.Id.checkoutBtn);
			checkoutButton.Click += (sender, e) => {
				if (CheckoutClicked != null)
					CheckoutClicked ();
			};

			shoppingCartView.PivotY = 0;
			shoppingCartView.PivotX = container.Width;

			return shoppingCartView;
		}

		public override void OnViewCreated (View view, Bundle savedInstanceState)
		{
			base.OnViewCreated (view, savedInstanceState);
			ListView.DividerHeight = 0;
			ListView.Divider = null;
			ListAdapter = new GroceryListAdapter (view.Context, _order.Products);
			if (ListAdapter.Count == 0)
				checkoutButton.Visibility = ViewStates.Invisible;
			_order.ProductsChanged += (sender, e) => {
				checkoutButton.Visibility = _order.Products.Any () ? ViewStates.Visible : ViewStates.Invisible;
			};
		}

		public Action CheckoutClicked { get; set; }
	}

	public class GroceryListAdapter : BaseAdapter
	{
		Context context;
		IReadOnlyList<Product> items;

		public GroceryListAdapter(Context context, IReadOnlyList<Product> items) : base() 
		{
			this.context = context;
			this.items = items;
		}

		public override long GetItemId (int position)
		{
			return position;
		}

		public override Java.Lang.Object GetItem (int position)
		{
			return new Java.Lang.String (items [position].ToString ());
		}

		public override int Count {
			get { return items.Count; }
		}

		public override View GetView (int position, View convertView, ViewGroup parent)
		{
			Product product = items [position];

			View view = convertView; // re-use an existing view, if one is available
			if (view == null) { // otherwise create a new one
				view = context
					.GetSystemService (Context.LayoutInflaterService)
					.JavaCast<LayoutInflater> ()
					.Inflate (Resource.Layout.BasketItem, parent, false);
				var swipper = ((SwipableListItem)view).SwipeListener;
				swipper.SwipeGestureBegin += (sender, e) => ((ListView)parent).RequestDisallowInterceptTouchEvent (true);
				swipper.SwipeGestureEnd += (sender, e) => ((ListView)parent).RequestDisallowInterceptTouchEvent (false);
				swipper.ItemSwipped += (sender, e) => {
					var p = ((ListView)parent).GetPositionForView (view);
					var order = WebService.Shared.CurrentOrder;
					order.Remove (order.Products [p]);
					NotifyDataSetChanged ();
				};
			}

			view.FindViewById<TextView> (Resource.Id.productTitle).Text = product.Name;
			view.FindViewById<TextView> (Resource.Id.productPrice).Text = product.PriceDescription;
			view.FindViewById<TextView> (Resource.Id.productColor).Text = product.Color.ToString();
			view.FindViewById<TextView> (Resource.Id.productSize).Text = product.Size.Description;

			var orderImage = view.FindViewById<ImageView> (Resource.Id.productImage);

			orderImage.SetImageResource (Resource.Drawable.blue_shirt);
			//No need to wait for the async download to return the view
			#pragma warning disable 4014
			orderImage.SetImageFromUrlAsync (product.ImageForSize (Images.ScreenWidth));
			#pragma warning restore 4014
			return view;
		}
	}
}

File: F#/XamarinStore.Droid/Fragments/BasketFragment.fs
namespace XamarinStore

open System
open System.Collections.Generic
open System.Globalization
open System.Linq
open System.Text
open Android.App
open Android.Content
open Android.OS
open Android.Runtime
open Android.Util
open Android.Views
open Android.Widget
open Order
open Product
open Images
open Helpers

type GroceryListAdapter(context:Context) =
    inherit Android.Widget.BaseAdapter()

    override this.GetItemId position =
        int64 position

    override this.GetItem position =
        new Java.Lang.String (WebService.CurrentOrder.Products.[position].ToString()) :> Java.Lang.Object

    override this.Count with get () = WebService.CurrentOrder.Products.Length

    override this.GetView (position, convertView, parent) =
        let product = WebService.CurrentOrder.Products.[position]

        let view = if not (convertView = null)
                   then convertView
                   else let tempView = LayoutInflater.From(context).Inflate (Resource_Layout.BasketItem, parent, false)
                        let swipper = (tempView :?> SwipableListItem).SwipeListener
                        swipper.SwipeGestureBegin <- fun () -> (parent :?> ListView).RequestDisallowInterceptTouchEvent true
                        swipper.SwipeGestureEnd <- fun () -> (parent :?> ListView).RequestDisallowInterceptTouchEvent false
                        swipper.ItemSwipped <- fun () -> if tempView.Parent <> null then
                                                             let p = (parent :?> ListView).GetPositionForView tempView
                                                             let productToRemove = WebService.CurrentOrder.Products.[p]
                                                             WebService.CurrentOrder <- WebService.CurrentOrder.WithoutProduct productToRemove
                                                             this.NotifyDataSetChanged ()
                        tempView

        view.FindViewById<TextView>(Resource_Id.productTitle).Text <- product.Name
        view.FindViewById<TextView>(Resource_Id.productPrice).Text <- product.GetPriceDescription()
        view.FindViewById<TextView>(Resource_Id.productColor).Text <- product.Color.ToString()
        view.FindViewById<TextView>(Resource_Id.productSize).Text <- product.Size.Description

        let orderImage = view.FindViewById<ImageView> (Resource_Id.productImage)
        orderImage.SetImageResource (Resource_Drawable.blue_shirt)
        //No need to wait for the async download to return the view
        Images.SetImageViewFromUrlAsync orderImage (product.ImageForSize (Images.ScreenWidth)) |> Async.StartImmediate
        view

type BasketFragment() =
    inherit ListFragment()

    let mutable checkoutButton = null

    override this.OnCreate savedInstanceState =
        base.OnCreate (savedInstanceState)
        this.RetainInstance <- true

    member val CheckoutClicked = fun ()->() with get, set

    override this.OnCreateView(inflater, container, savedInstanceState) =
        let shoppingCartView = inflater.Inflate (Resource_Layout.Basket, container, false)

        checkoutButton <- shoppingCartView.FindViewById<Button> (Resource_Id.checkoutBtn)
        checkoutButton.Click.Add(fun x-> this.CheckoutClicked ())

        shoppingCartView.PivotY <- 0.0f
        shoppingCartView.PivotX <- float32 container.Width

        shoppingCartView
    
    override this.OnViewCreated (view, savedInstanceState) =
        base.OnViewCreated (view, savedInstanceState)
        this.ListView.DividerHeight <- 0
        this.ListView.Divider <- null
        this.ListAdapter <- (new GroceryListAdapter (view.Context) :> IListAdapter)
        if this.ListAdapter.Count = 0 then
            checkoutButton.Visibility <- ViewStates.Invisible
        WebService.CurrentOrder <- { WebService.CurrentOrder 
                                        with Notification = Some (fun order -> checkoutButton.Visibility <- if order.Products.Any () 
                                                                                                                then ViewStates.Visible 
                                                                                                                else ViewStates.Invisible) }